### PR TITLE
refactor(curriculum): improve binary search tree

### LINKED ISSUE #52586: [python]: learn-tree-traversal-by-building-a-binary-search-tree
**TO FIX POST BETA RELEASE**

```markdown
- <step-number>: <bug description>
```

```[tasklist]
### Tasks
- [x] The methods with the `_` are being used before being defined. This should be refactored.
- [x] Check if user has removed pass or not when asked.
```

### PR DESCRIPTION
**Pull request by zairahira**
Checklist:

- [x] I have read and followed the [contribution guidelines](https://contribute.freecodecamp.org).
- [x] I have read and followed the [how to open a pull request guide](https://contribute.freecodecamp.org/#/how-to-open-a-pull-request).
- [x] My pull request targets the `main` branch of freeCodeCamp.
- [x] I have tested these changes either locally on my machine, or GitPod.

Closes #52586

### CODE REVIEW

File: curriculum/challenges/english/07-scientific-computing-with-python/learn-tree-traversal-by-building-a-binary-search-tree/65c644829cfb63acf3479d09.md [hunk 1]
```diff
@@ -1,63 +1,86 @@
...
+A list of `nodes` is given for you to insert into the `BinarySearchTree` instance `bst`.
 
-Within the `BinarySearchTree` class, define a `delete` method. It takes two parameters: `self` and `key`.
+You can insert multiple nodes into the tree like this:
 
-`key` is the value that the user wants to delete from the binary search tree.
+```python
+samples = [1, 2, 3, 4, 5]
+
+for sample in samples:
```
Comment (Dario-DC) [inline curriculum/challenges/english/07-scientific-computing-with-python/learn-tree-traversal-by-building-a-binary-search-tree/65c644829cfb63acf3479d09.md:17]: It should be either `for node`  or `your_object.insert(sample)` to work. Anyway, I don't think the example is necessary. You can ask the camper to write a for loop to iterate over the list and call `self.insert` for each item.

File: curriculum/challenges/english/07-scientific-computing-with-python/learn-tree-traversal-by-building-a-binary-search-tree/65c4f33bc3c3d8db7f732787.md [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+dashedName: step-16
+---
+
+# --description--
+
+Now, to perform the actual insertion, define an empty `insert` method within the `BinarySearchTree` class and give it a `self` parameter. Remember to use the `pass` keyword.
+
+# --hints--
+
+You should define an `insert` method with a `self` parameter within the `BinarySearchTree` class. Remember the `pass` keyword.
```
Comment (ilenia-magoni) [inline curriculum/challenges/english/07-scientific-computing-with-python/learn-tree-traversal-by-building-a-binary-search-tree/65c4f33bc3c3d8db7f732787.md:14]: this hint says pretty much the same thing as the instructions, an hint should be sufficently different to give an extra help

File: curriculum/challenges/english/07-scientific-computing-with-python/learn-tree-traversal-by-building-a-binary-search-tree/65c645b838e7deb080fc25e0.md [hunk 3]
```diff
@@ -1,62 +1,87 @@
...
 
-The deletion operation might result in a new root (for example, if the node to be deleted is the current root).
+Now, the nodes have been inserted. To see if they have  been correctly inserted, you can search for a node in the tree.
 
-To handle this case, assign the result of the `_delete` call to `self.root`.
+You can search for a node in the tree like this:
+
+```python
+your_object.search(node)
+```
```
Comment (ilenia-magoni) [inline curriculum/challenges/english/07-scientific-computing-with-python/learn-tree-traversal-by-building-a-binary-search-tree/65c645b838e7deb080fc25e0.md:16]: This is step 28 of the 10th project, is it really necessary to remind the camper how to call a method?

File: curriculum/challenges/english/07-scientific-computing-with-python/learn-tree-traversal-by-building-a-binary-search-tree/65c646d4148ae3b2d1cbcac4.md [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+id: 65c646d4148ae3b2d1cbcac4
+title: Step 29
+challengeType: 20
+dashedName: step-29
+---
+
+# --description--
+
+Note that, your search returns something like `80: <__main__.TreeNode object at 0x108b3e0>`. This is the object.
+This is default string representation when printing an instance of a class.
```
Comment (ilenia-magoni) [inline curriculum/challenges/english/07-scientific-computing-with-python/learn-tree-traversal-by-building-a-binary-search-tree/65c646d4148ae3b2d1cbcac4.md:11]:
```suggestion
This is the default string representation when printing an instance of a class.
```

File: curriculum/challenges/english/07-scientific-computing-with-python/learn-tree-traversal-by-building-a-binary-search-tree/65c9de201959f73591b606e6.md [hunk 5]
```diff
@@ -0,0 +1,125 @@
...
+id: 65c9de201959f73591b606e6
+title: Step 31
+challengeType: 20
+dashedName: step-31
+---
+
+# --description--
+
+As you can see, now you get a readable output in the console.
+Now, remove the code.
```
Comment (ilenia-magoni) [inline curriculum/challenges/english/07-scientific-computing-with-python/learn-tree-traversal-by-building-a-binary-search-tree/65c9de201959f73591b606e6.md:11]: which code?

File: curriculum/challenges/english/07-scientific-computing-with-python/learn-tree-traversal-by-building-a-binary-search-tree/65ca298b039a852d149f2e95.md [hunk 6]
```diff
@@ -1,8 +1,8 @@
 ---
-id: 655a580da8b2419496c88f61
-title: Step 56
+id: 65ca298b039a852d149f2e95
+title: Step 60
 challengeType: 20
-dashedName: step-56
+dashedName: step-60
 ---
```
Comment (Dario-DC) [inline curriculum/challenges/english/07-scientific-computing-with-python/learn-tree-traversal-by-building-a-binary-search-tree/65ca298b039a852d149f2e95.md:7]: In this step the `print("Search for 40:", bst.search(40))` call can be uncommented.

File: curriculum/challenges/english/07-scientific-computing-with-python/learn-tree-traversal-by-building-a-binary-search-tree/65ca06475e9b8aff8ea5a5f1.md [hunk 7]
```diff
@@ -1,13 +1,13 @@
...
+title: Step 34
 challengeType: 20
-dashedName: step-31
+dashedName: step-34
 ---
 
 # --description--
 
-After your existing `if`, write another one that checks if the target key is less than the current node key.
+After your existing `if`, write another one that checks if the target key is less than the current node key. Remember to use `pass`.
```
Comment (ilenia-magoni) [inline curriculum/challenges/english/07-scientific-computing-with-python/learn-tree-traversal-by-building-a-binary-search-tree/65ca06475e9b8aff8ea5a5f1.md:10]: At some point we should be able to avoid this reminder, and maybe put it as an hint if necessary

**Review by majestic-owl448 — CHANGES_REQUESTED** [review]

**Review by Dario-DC — COMMENTED** [review]
I went through the steps quickly. In general, we should probably check that `pass` is removed when we asked to do it. Also check the right indentation for the code required for each step.

File: client/src/pages/learn/scientific-computing-with-python/refactor-bst/index.md
Comment (naomi-lgbt) [inline client/src/pages/learn/scientific-computing-with-python/refactor-bst/index.md:1]: This file can be deleted, the temporary block is gone.

File: client/i18n/locales/english/intro.json [hunk 8]
```diff
@@ -828,7 +828,8 @@
         "intro": [
           "It's time to put your Python skills to the test. By completing these projects, you'll demonstrate that you have a strong foundational knowledge of Python. And you'll qualify for freeCodeCamp's Scientific Computing with Python Certification."
         ]
-      }
+      },
+      "refactor-bst": { "title": "refactor-bst", "intro": ["", ""] }
```
Comment (naomi-lgbt) [inline client/i18n/locales/english/intro.json:832]:
```suggestion
      }
```

**Review by naomi-lgbt — CHANGES_REQUESTED** [review]

File: curriculum/challenges/english/07-scientific-computing-with-python/learn-tree-traversal-by-building-a-binary-search-tree/65ca06475e9b8aff8ea5a5f1.md (hunk 7 above)
Comment (zairahira) [reply curriculum/challenges/english/07-scientific-computing-with-python/learn-tree-traversal-by-building-a-binary-search-tree/65ca06475e9b8aff8ea5a5f1.md]: Such instructions are now in the hint only. This is an advanced project and we can skip telling the campers to do this in every instruction.

File: curriculum/challenges/english/07-scientific-computing-with-python/learn-tree-traversal-by-building-a-binary-search-tree/65c4eb814cc977c95cd7df0e.md [hunk 9]
```diff
@@ -1,12 +1,13 @@
 ---
-id: 65560f9380be92226084ef46
+id: 65c4eb814cc977c95cd7df0e
```
[the remark this answers is not in the document]
Comment (zairahira) [reply curriculum/challenges/english/07-scientific-computing-with-python/learn-tree-traversal-by-building-a-binary-search-tree/65c4eb814cc977c95cd7df0e.md:2]: The project has almost all of the files reordered and shuffled. New steps have also been added. IMO, it is better for the campers to restart the project, provided the restructure being this big.

File: curriculum/challenges/english/07-scientific-computing-with-python/learn-tree-traversal-by-building-a-binary-search-tree/65c64057a080baa7d60523ed.md [hunk 10]
```diff
@@ -1,8 +1,8 @@
 ---
-id: 655a16db1ccc5e132b5bc44d
-title: Step 21
+id: 65c64057a080baa7d60523ed
+title: Step 25
```
Comment (ilenia-magoni) [inline curriculum/challenges/english/07-scientific-computing-with-python/learn-tree-traversal-by-building-a-binary-search-tree/65c64057a080baa7d60523ed.md:3]: can this step be merged into step 24? or step 24 becomes too complex then?

File: curriculum/challenges/english/07-scientific-computing-with-python/learn-tree-traversal-by-building-a-binary-search-tree/65c644829cfb63acf3479d09.md [hunk 11]
```diff
@@ -1,61 +1,77 @@
 ---
-id: 655a45c52fa3ea5ece3034c7
+id: 65c644829cfb63acf3479d09
 title: Step 27
```
Comment (ilenia-magoni) [inline curriculum/challenges/english/07-scientific-computing-with-python/learn-tree-traversal-by-building-a-binary-search-tree/65c644829cfb63acf3479d09.md:3]: why there is code appearing magically from nowhere here? it doesn't look correct... we should give it in the instructions and let the camper put it in the editor

File: curriculum/challenges/english/07-scientific-computing-with-python/learn-tree-traversal-by-building-a-binary-search-tree/65c644829cfb63acf3479d09.md [hunk 12]
```diff
@@ -1,61 +1,77 @@
...
 
-Inside the `delete` method, delete `pass` and call the private helper method `_delete` with the root of the binary search tree (BST) and the key to delete as the arguments.
+A list of `nodes` is given for you to insert into the `BinarySearchTree` instance `bst`.
+
+Insert the nodes in the list `nodes` into the `bst` instance by iterating over the list and calling the `insert` method on each node in the list.
 
 # --hints--
 
-Your `delete` method should call the `_delete` method with two arguments: `self.root` and `key`.
+You should have a `for node in nodes` loop.
```
Comment (ilenia-magoni) [inline curriculum/challenges/english/07-scientific-computing-with-python/learn-tree-traversal-by-building-a-binary-search-tree/65c644829cfb63acf3479d09.md:16]: It should be explicit in the instructions that we want a `node` variable here

File: curriculum/challenges/english/07-scientific-computing-with-python/learn-tree-traversal-by-building-a-binary-search-tree/65ca05f7cba1e6fe70527534.md [hunk 13]
```diff
@@ -1,8 +1,8 @@
 ---
-id: 655a46fce0ce5a638c180e36
-title: Step 30
+id: 65ca05f7cba1e6fe70527534
```
Comment (ilenia-magoni) [inline curriculum/challenges/english/07-scientific-computing-with-python/learn-tree-traversal-by-building-a-binary-search-tree/65ca05f7cba1e6fe70527534.md:2]: the two description paragraphs could be divided differently. Maybe like

```
Inside your `_delete` method, replace `pass` with an `if` statement that checks if the current `node` is `None`.

When the current node is `None`, the key to be deleted was not found. Therefore, return `node` from your `if` block.
```

File: curriculum/challenges/english/07-scientific-computing-with-python/learn-tree-traversal-by-building-a-binary-search-tree/65ca06864129380054b55dc0.md [hunk 14]
```diff
@@ -1,8 +1,8 @@
 ---
-id: 655a47a9404d856743c7f529
-title: Step 32
+id: 65ca06864129380054b55dc0
```
Comment (ilenia-magoni) [inline curriculum/challenges/english/07-scientific-computing-with-python/learn-tree-traversal-by-building-a-binary-search-tree/65ca06864129380054b55dc0.md:2]: This step could be merged into the previous step

File: curriculum/challenges/english/07-scientific-computing-with-python/learn-tree-traversal-by-building-a-binary-search-tree/65ca0a1f27596a089b0363b9.md [hunk 15]
```diff
@@ -1,104 +1,112 @@
...
+title: Step 48
 challengeType: 20
-dashedName: step-54
+dashedName: step-48
 ---
 
 # --description--
 
-Write a `for` loop to iterate over the `nodes` list. Inside the `for` loop body, call the `insert` method of the `bst` object, passing the node at the current iteration to insert all values orderly into the binary search tree.
+Inside the `delete` method, delete `pass` and call the private helper method `_delete` with the root of the binary search tree (BST) and the key to delete as the arguments.
```
Comment (ilenia-magoni) [inline curriculum/challenges/english/07-scientific-computing-with-python/learn-tree-traversal-by-building-a-binary-search-tree/65ca0a1f27596a089b0363b9.md:10]: why do we need to specify BST now?

File: curriculum/challenges/english/07-scientific-computing-with-python/learn-tree-traversal-by-building-a-binary-search-tree/65cb40ee75b4f787b21066a8.md [hunk 16]
```diff
@@ -1,87 +1,105 @@
 ---
-id: 655a5489c62889851c60ff4b
-title: Step 45
+id: 65cb40ee75b4f787b21066a8
```
Comment (ilenia-magoni) [inline curriculum/challenges/english/07-scientific-computing-with-python/learn-tree-traversal-by-building-a-binary-search-tree/65cb40ee75b4f787b21066a8.md:2]: this step can probably be merged to the previous one

File: curriculum/challenges/english/07-scientific-computing-with-python/learn-tree-traversal-by-building-a-binary-search-tree/65ca2a52d579b22feb89177f.md [hunk 17]
```diff
@@ -1,13 +1,13 @@
...
+title: Step 62
 challengeType: 20
-dashedName: step-61
+dashedName: step-62
 ---
 
 # --description--
 
-As a last step, search for `40` again. It should be deleted. You can copy and paste the previous `print()` call.
+As a last step, search for `40` again. It should be deleted.
```
Comment (ilenia-magoni) [inline curriculum/challenges/english/07-scientific-computing-with-python/learn-tree-traversal-by-building-a-binary-search-tree/65ca2a52d579b22feb89177f.md:10]: why again?
```suggestion
As a last step, search for `40`. It should be deleted.
```

**Review by majestic-owl448 — CHANGES_REQUESTED** [review]
I thinkl we are almost there, I have left a few comments.

**Review by Dario-DC — CHANGES_REQUESTED** [review]
Reviewed until step 40

File: curriculum/challenges/english/07-scientific-computing-with-python/learn-tree-traversal-by-building-a-binary-search-tree/65ca294e05dc032bf8922dc7.md [hunk 18]
```diff
@@ -1,13 +1,13 @@
...
+title: Step 59
 challengeType: 20
-dashedName: step-55
+dashedName: step-59
 ---
 
 # --description--
 
-Below the loop, call the `print()` function and pass it the string `Inorder traversal:` as the first argument and the `bst.inorder_traversal()` call as the second argument to print the result of the inorder traversal.
+Below the print statement, call the `print()` function and pass it the string `Inorder traversal:` as the first argument and the `bst.inorder_traversal()` call as the second argument to print the result of the inorder traversal.
```
Comment (Dario-DC) [inline curriculum/challenges/english/07-scientific-computing-with-python/learn-tree-traversal-by-building-a-binary-search-tree/65ca294e05dc032bf8922dc7.md:10]:
```suggestion
Below the `print()` call, call again `print()` and pass it the string `'Inorder traversal:'` as the first argument and the `bst.inorder_traversal()` call as the second argument to print the result of the inorder traversal.
```

File: curriculum/challenges/english/07-scientific-computing-with-python/learn-tree-traversal-by-building-a-binary-search-tree/65ca2a18039c942f04ddde83.md [hunk 19]
```diff
@@ -1,8 +1,8 @@
 ---
-id: 655a5908085cdf99b7630646
-title: Step 60
+id: 65ca2a18039c942f04ddde83
+title: Step 61
 challengeType: 20
-dashedName: step-60
+dashedName: step-61
 ---
```
Comment (Dario-DC) [inline curriculum/challenges/english/07-scientific-computing-with-python/learn-tree-traversal-by-building-a-binary-search-tree/65ca2a18039c942f04ddde83.md:7]: Quotes should be added to the strings in description/hints.

File: curriculum/challenges/english/07-scientific-computing-with-python/learn-tree-traversal-by-building-a-binary-search-tree/65ca2a52d579b22feb89177f.md [hunk 20]
```diff
@@ -1,13 +1,13 @@
 ---
-id: 655a59607b0d2e9b0f5d69e5
-title: Step 61
+id: 65ca2a52d579b22feb89177f
+title: Step 62
 challengeType: 20
-dashedName: step-61
+dashedName: step-62
 ---
```
Comment (Dario-DC) [inline curriculum/challenges/english/07-scientific-computing-with-python/learn-tree-traversal-by-building-a-binary-search-tree/65ca2a52d579b22feb89177f.md:7]: I would swap the last two steps. So we can test if `delete` has deleted the key `40` and then see the inorder traversal as the last step.

**Review by Dario-DC — COMMENTED** [review]
Left few other comments. I think we are almost there.

File: curriculum/challenges/english/07-scientific-computing-with-python/learn-tree-traversal-by-building-a-binary-search-tree/65c4f22498d22ed775ef8efb.md [hunk 21]
```diff
@@ -1,13 +1,15 @@
...
+title: Step 10
 challengeType: 20
-dashedName: step-13
+dashedName: step-10
 ---
 
 # --description--
 
-Now you need to check if the `node` parameter is `None`. If it is, this means that the method has reached a leaf node or an empty spot in the tree where the new node should be inserted.
+Now you need to check if the `node` parameter is `None`(translates to `if node is None`). If it is, this means that the method has reached a leaf node or an empty spot in the tree where the new node should be inserted.
```
[the remark this answers is not in the document]
Comment (zairahira) [reply curriculum/challenges/english/07-scientific-computing-with-python/learn-tree-traversal-by-building-a-binary-search-tree/65c4f22498d22ed775ef8efb.md:10]: I'll go with the second one.

File: curriculum/challenges/english/07-scientific-computing-with-python/learn-tree-traversal-by-building-a-binary-search-tree/65c64057a080baa7d60523ed.md (hunk 10 above)
Comment (zairahira) [reply curriculum/challenges/english/07-scientific-computing-with-python/learn-tree-traversal-by-building-a-binary-search-tree/65c64057a080baa7d60523ed.md:3]: Step-24 would be too wordy if we merge it with 25.

File: curriculum/challenges/english/07-scientific-computing-with-python/learn-tree-traversal-by-building-a-binary-search-tree/65c644829cfb63acf3479d09.md (hunk 11 above)
Comment (zairahira) [reply curriculum/challenges/english/07-scientific-computing-with-python/learn-tree-traversal-by-building-a-binary-search-tree/65c644829cfb63acf3479d09.md:3]: i have added another step for creating a list.

File: curriculum/challenges/english/07-scientific-computing-with-python/learn-tree-traversal-by-building-a-binary-search-tree/65cb40ee75b4f787b21066a8.md (hunk 16 above)
Comment (zairahira) [reply curriculum/challenges/english/07-scientific-computing-with-python/learn-tree-traversal-by-building-a-binary-search-tree/65cb40ee75b4f787b21066a8.md:2]: 49/50 combined now.

File: curriculum/challenges/english/07-scientific-computing-with-python/learn-tree-traversal-by-building-a-binary-search-tree/65ca0e2d05557a13a40b1243.md [hunk 22]
```diff
@@ -1,8 +1,8 @@
 ---
-id: 655a577302a8a791ed62e8d9
-title: Step 51
+id: 65ca0e2d05557a13a40b1243
+title: Step 53
 challengeType: 20
-dashedName: step-51
+dashedName: step-53
 ---
```
[the remark this answers is not in the document]
Comment (zairahira) [reply curriculum/challenges/english/07-scientific-computing-with-python/learn-tree-traversal-by-building-a-binary-search-tree/65ca0e2d05557a13a40b1243.md:7]: Fixed.

File: curriculum/challenges/english/07-scientific-computing-with-python/learn-tree-traversal-by-building-a-binary-search-tree/65ca071d5921760254949f76.md [hunk 23]
```diff
@@ -1,8 +1,8 @@
 ---
-id: 655a489d83b1996bd537b153
-title: Step 34
+id: 65ca071d5921760254949f76
+title: Step 37
 challengeType: 20
-dashedName: step-34
+dashedName: step-37
 ---
```
[the remark this answers is not in the document]
Comment (zairahira) [reply curriculum/challenges/english/07-scientific-computing-with-python/learn-tree-traversal-by-building-a-binary-search-tree/65ca071d5921760254949f76.md:7]: done!

File: curriculum/challenges/english/07-scientific-computing-with-python/learn-tree-traversal-by-building-a-binary-search-tree/65ca2a52d579b22feb89177f.md (hunk 20 above)
Comment (zairahira) [reply curriculum/challenges/english/07-scientific-computing-with-python/learn-tree-traversal-by-building-a-binary-search-tree/65ca2a52d579b22feb89177f.md:7]: From what I understood, this should be the flow?

1. Delete `40`
2. Search `40`
3. Print tree after deleting `40`

Currently it is: delete, print, search.

File: curriculum/challenges/english/07-scientific-computing-with-python/learn-tree-traversal-by-building-a-binary-search-tree/65ca0794ec3ed103bca67ed7.md [hunk 24]
```diff
@@ -1,8 +1,8 @@
 ---
-id: 655a493ab909a96f7c316cd5
-title: Step 36
+id: 65ca0794ec3ed103bca67ed7
+title: Step 39
 challengeType: 20
-dashedName: step-36
+dashedName: step-39
 ---
```
[the remark this answers is not in the document]
Comment (zairahira) [reply curriculum/challenges/english/07-scientific-computing-with-python/learn-tree-traversal-by-building-a-binary-search-tree/65ca0794ec3ed103bca67ed7.md:7]: I have added the test for this step.

File: curriculum/challenges/english/07-scientific-computing-with-python/learn-tree-traversal-by-building-a-binary-search-tree/65ca2a52d579b22feb89177f.md (hunk 20 above)
Comment (Dario-DC) [reply curriculum/challenges/english/07-scientific-computing-with-python/learn-tree-traversal-by-building-a-binary-search-tree/65ca2a52d579b22feb89177f.md:7]: Yes, this is what I meant.

**Review by Dario-DC — COMMENTED** [review]
Works perfectly! Left a few comments on small things.

File: curriculum/challenges/english/07-scientific-computing-with-python/learn-tree-traversal-by-building-a-binary-search-tree/65c4f2b7178afed8e88f782f.md [hunk 25]
```diff
@@ -1,18 +1,32 @@
...
+id: 65c4f2b7178afed8e88f782f
+title: Step 12
 challengeType: 20
-dashedName: step-15
+dashedName: step-12
 ---
 
 # --description--
 
 If `key` < `node.key` returns `True`, then the new node should be placed in the left subtree.
```
Comment (ilenia-magoni) [inline curriculum/challenges/english/07-scientific-computing-with-python/learn-tree-traversal-by-building-a-binary-search-tree/65c4f2b7178afed8e88f782f.md:10]: I would say "evaluates to `True`" instead of "returns"

**Review by majestic-owl448 — COMMENTED** [review]
I have one little nitpick, then I think it's good to go

**Review by majestic-owl448 — APPROVED** [review]
LGTM 👍

File: curriculum/challenges/english/07-scientific-computing-with-python/learn-tree-traversal-by-building-a-binary-search-tree/65c63fd9b387f0a6c5bf6a72.md [hunk 26]
```diff
@@ -1,13 +1,13 @@
...
+dashedName: step-24
 ---
 
 # --description--
 
-Inside the `search` method, delete `pass` and call the private helper method `_search` with the following arguments. You will define `_search` in the next steps.
+Inside the `search` method, delete `pass` and call the helper method `_search` with the following arguments.
 
 - `self.root`: This is the root of the binary search tree. The search starts from the root.
 - `key`: This is the value that the user wants to find in the binary search tree.
```
Comment (Dario-DC) [inline curriculum/challenges/english/07-scientific-computing-with-python/learn-tree-traversal-by-building-a-binary-search-tree/65c63fd9b387f0a6c5bf6a72.md:13]: Not able to comment in the right place. Anyway, I would remove that "private" from below, since the `_` does not make the method really private. "helper" seems more appropriate and it has been already used before.

File: curriculum/challenges/english/07-scientific-computing-with-python/learn-tree-traversal-by-building-a-binary-search-tree/65c6409418806da8d0636ffc.md [hunk 27]
```diff
@@ -0,0 +1,84 @@
...
+title: Step 26
+challengeType: 20
+dashedName: step-26
+---
+
+# --description--
+
+The `insert` and `search` functionalities are complete, it's time to test them.
+
+Note that, at this point, the nodes are not sorted and just inserted. We will work on sorting using inorder traversal later on.
```
Comment (Dario-DC) [inline curriculum/challenges/english/07-scientific-computing-with-python/learn-tree-traversal-by-building-a-binary-search-tree/65c6409418806da8d0636ffc.md:12]:
```suggestion
Note that, at this point, the nodes are not sorted and just inserted. You'll work on sorting using inorder traversal later on.
```

File: curriculum/challenges/english/07-scientific-computing-with-python/learn-tree-traversal-by-building-a-binary-search-tree/65ca0a1f27596a089b0363b9.md [hunk 28]
```diff
@@ -1,28 +1,47 @@
...
+title: Step 49
 challengeType: 20
-dashedName: step-60
+dashedName: step-49
 ---
 
 # --description--
 
-Corfirm that `40` has been deleted by printing the tree. After deleting `40`, call `print()` passing the string `Inorder traversal after deleting 40:` as the first argument and an `inorder_traversal()` call as the second argument.
+Inside the `delete` method, delete `pass` and call the private helper method `_delete` with the root of the Binary Search Tree and the key to delete as the arguments.
```
Comment (Dario-DC) [inline curriculum/challenges/english/07-scientific-computing-with-python/learn-tree-traversal-by-building-a-binary-search-tree/65ca0a1f27596a089b0363b9.md:10]:
```suggestion
Inside the `delete` method, delete `pass` and call the helper method `_delete` with the root of the Binary Search Tree and the key to delete as the arguments.
```

File: curriculum/challenges/english/07-scientific-computing-with-python/learn-tree-traversal-by-building-a-binary-search-tree/65ca0d5adf39c410cd1177cc.md [hunk 29]
```diff
@@ -1,104 +1,107 @@
...
+title: Step 50
 challengeType: 20
-dashedName: step-56
+dashedName: step-50
 ---
 
 # --description--
 
-Below your `print()` call, add another `print()` call to test the search functionality. This time, pass the string `Search for 40:` as the first argument. For the second argument, call the `search` method of `bst` and pass `40` as the argument.
+Now, you would work on traversing the tree based on the in-order traversal method. In-order traversal is a depth-first binary tree traversal algorithm that visits the left subtree, the current node, and then the right subtree.
```
Comment (Dario-DC) [inline curriculum/challenges/english/07-scientific-computing-with-python/learn-tree-traversal-by-building-a-binary-search-tree/65ca0d5adf39c410cd1177cc.md:10]:
```suggestion
Now, you'll work on traversing the tree based on the in-order traversal method. In-order traversal is a depth-first binary tree traversal algorithm that visits the left subtree, the current node, and then the right subtree.
```

File: curriculum/challenges/english/07-scientific-computing-with-python/learn-tree-traversal-by-building-a-binary-search-tree/65ca0e8eb2c9c215269d6a66.md [hunk 30]
```diff
@@ -1,15 +1,13 @@
...
-dashedName: step-44
+dashedName: step-54
 ---
 
 # --description--
 
-The `inorder_traversal` method is responsible for performing an in-order traversal of the binary search tree. It returns the keys of the nodes in sorted order.
-
-In-order traversal is a depth-first binary tree traversal algorithm that visits the left subtree, the current node, and then the right subtree.
+The `inorder_traversal` method is responsible for performing an in-order traversal of the binary search tree. It returns the keys of the nodes in sorted order. It would use the helper method `_inorder_traversal` to achieve this.
```
Comment (Dario-DC) [inline curriculum/challenges/english/07-scientific-computing-with-python/learn-tree-traversal-by-building-a-binary-search-tree/65ca0e8eb2c9c215269d6a66.md:10]:
```suggestion
The `inorder_traversal` method is responsible for performing an in-order traversal of the binary search tree. It returns the keys of the nodes in sorted order. It will use the helper method `_inorder_traversal` to achieve this.
```

File: curriculum/challenges/english/07-scientific-computing-with-python/learn-tree-traversal-by-building-a-binary-search-tree/65ca294e05dc032bf8922dc7.md [hunk 31]
```diff
@@ -1,17 +1,17 @@
...
+title: Step 59
 challengeType: 20
-dashedName: step-55
+dashedName: step-59
 ---
 
 # --description--
 
-Below the loop, call the `print()` function and pass it the string `Inorder traversal:` as the first argument and the `bst.inorder_traversal()` call as the second argument to print the result of the inorder traversal.
+Below the print statement, call the `print()` function and pass it the string `'Inorder traversal:'` as the first argument and the `bst.inorder_traversal()` call as the second argument to print the result of the inorder traversal.
```
Comment (Dario-DC) [inline curriculum/challenges/english/07-scientific-computing-with-python/learn-tree-traversal-by-building-a-binary-search-tree/65ca294e05dc032bf8922dc7.md:10]:
```suggestion
Below the `print()` call, call the `print()` function again and pass it the string `'Inorder traversal:'` as the first argument and the `bst.inorder_traversal()` call as the second argument to print the result of the inorder traversal.
```

File: curriculum/challenges/english/07-scientific-computing-with-python/learn-tree-traversal-by-building-a-binary-search-tree/65ca2a18039c942f04ddde83.md [hunk 32]
```diff
@@ -1,108 +1,119 @@
...
 
 # --hints--
 
-You should return the string value of `self.key` from your `__str__` method.
+You should have `print('Search for 40:', bst.search(40))` in your code.
+
 
 ```js
-({ test: () => assert.match(code, /^\s{8}return\s+str\(\s*self\.key\s*\)/m) })
+({ test: () => assert.match(code, /^print\(("|')Search for 40:("|'), bst\.search\(40\)\)/m) })
```
Comment (Dario-DC) [inline curriculum/challenges/english/07-scientific-computing-with-python/learn-tree-traversal-by-building-a-binary-search-tree/65ca2a18039c942f04ddde83.md:20]:
```suggestion
({ test: () => assert.match(code, /^print\(\s*("|')Search for 40:("|')\s*,\s*bst\.search\(\s*40\s*\)\s*\)/m) })
```

**Review by Dario-DC — COMMENTED** [review]
Just a few nitpicks. After that it's ready to go :rocket:

**Review by Dario-DC — APPROVED** [review]

**Comment by Sembauke** [comment]
Hey @naomi-lgbt,
Do you have a moment to review for the Python team?

File: curriculum/challenges/english/07-scientific-computing-with-python/learn-tree-traversal-by-building-a-binary-search-tree/65c9ddd336596e30a4266a50.md [hunk 33]
```diff
@@ -0,0 +1,93 @@
+---
+id: 65c9ddd336596e30a4266a50
+title: Step 31
+challengeType: 20
+dashedName: step-31
+---
+
+# --description--
+
+In the body of the `__str__` method, delete `pass` and return the result of calling the `str()` function with `self.key` as the argument.  This is the attribute of the current node object that stores the value associated with the node.
```
Comment (zairahira) [inline curriculum/challenges/english/07-scientific-computing-with-python/learn-tree-traversal-by-building-a-binary-search-tree/65c9ddd336596e30a4266a50.md:10]: Reminder: Instruction should be rephrased as the `__str__` method is already discussed in detail in the refactored sudoku solver.

**Review by naomi-lgbt — APPROVED** [review]
